### ISSUE TITLE
`TaskGroup.nbytes_in_memory` miscounted for replicated keys

### ISSUE THREAD

**Issue by gjoseph92** [dask/distributed#4927]
I think there is a logic error with bookkeping for `TaskGroup.nbytes_in_memory`. There's a discrepancy between how we increment it and decrement it when multiple workers hold the same key.

In `transition_memory_released`, we decrement it by `nbytes` once for every worker that holds that task: https://github.com/dask/distributed/blob/6340b5b34a582ff1ffe3eaf9dd7f3aac33d0a983/distributed/scheduler.py#L2646-L2649
Whereas in `_propagate_forgotten`, we decrement it once by `nbytes` if there are any workers holding the task, regardless of how many. This doesn't match with `transition_memory_released`:
https://github.com/dask/distributed/blob/646b12be37ef82e2afa2baf2827716746fa93173/distributed/scheduler.py#L7339-L7341

On the creation side, in `TaskState.set_nbytes`, we only increment it by the _diff_ between the last known value and the current value. If the key is being copied to multiple workers, this difference is usually 0: https://github.com/dask/distributed/blob/6340b5b34a582ff1ffe3eaf9dd7f3aac33d0a983/distributed/scheduler.py#L1556-L1562

In short, I think `TaskGroup.nbytes_in_memory` is incremented once per key, but decremented once per _copy_ of the key.

If `nbytes` can be different for different workers, then to do this bookkeeping correctly, I think we'd also need to track `TaskState.total_nbytes` (size of all copies of the key), then decrement by that once in `transition_memory_released` and `_propagate_forgotten`.

Discovered in https://github.com/dask/distributed/pull/4925#issuecomment-863600812. I think #4925 made this more apparent, since it encourages more data replication.

cc @crusaderky since you know more about replicated keys.

**Comment by fjetter** [comment]
I stumbled over this myself recently, see https://github.com/dask/distributed/blob/fced981adf6b8d95900cc3e089237201251f803f/distributed/tests/test_scheduler.py#L1919-L1920 where this behaviour is intentionally pinned but I believe this was by mistake. Intuitively, I would also expect this to be different.

From what I can see, the lack of counting is introduced in `Scheduler.add_keys` where workers let the scheduler know once they have a replica. There, the group nbytes is not increased
https://github.com/dask/distributed/blob/fced981adf6b8d95900cc3e089237201251f803f/distributed/scheduler.py#L6300-L6303 

I noticed this in my deadlock PR which already grew without bounds so I didn't fix it.

**Comment by mrocklin** [comment]
Maybe this will solve the problem? https://github.com/dask/distributed/pull/4930

**Comment by jrbourbeau** [comment]
Closed via https://github.com/dask/distributed/pull/4930
